app/src/main/java/com/github/andrezimmermann/gluappsample/server/api/command/RestConnection.java

```diff
@@ -0,0 +1,249 @@
+package com.github.andrezimmermann.gluappsample.server.api.command;
+
+
+import com.github.andrezimmermann.gluappsample.server.api.Endpoint;
+import com.github.andrezimmermann.gluappsample.server.converter.DataConverter;
+import com.github.andrezimmermann.gluappsample.server.data.RequestData;
+import com.github.andrezimmermann.gluappsample.server.data.ResponseData;
+import com.github.andrezimmermann.gluappsample.shared.error.ServiceException;
+
+import org.apache.commons.codec.binary.Base64;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStreamWriter;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.net.HttpURLConnection;
+import java.net.ProtocolException;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Wraps the HttpConnection to the WebService
+ *
+ * @param <T> the kind of data it process
+ */
+public class RestConnection<T extends RequestData<R>, R extends ResponseData> {
+
+
+    private DataConverter converter;
+    private String rootUrl = "https://api.appglu.com/v1/queries/";
+    private HttpRequest httpRequest;
+    private Class<R> resultClass;
+
+    RestConnection(Endpoint endpoint, DataConverter converter) {
+        this.converter = converter;
+        httpRequest = new HttpRequest(rootUrl + endpoint.getLocation()).
+                basicAuth("WKD4N7YMA1uiM8V", "DtdTtzMLQlA0hk2C1Yi5pLyVIlAQ68").header("X-AppGlu-Environment", "staging");
+
+        resultClass = findTypeParameters(getClass(), RestConnection.class)[1];
+    }
+
+    //Reflection helper, to get type parameters
+    public <S, B extends S> Class[] findTypeParameters(Class<B> base, Class<S> superClass) {
```

Comment:
Why you had to use reflection? What is the tradeoff of using it on Android? How would you avoid it?